data/src/main/java/com/linkedin/data/schema/PdlBuilder.java

```diff
@@ -275,7 +298,36 @@ else if (value instanceof DataList)
     {
       return jsonCodec.listToString((DataList) value);
     }
-    else if (value instanceof String)
+
+    return toPrimitiveValue(value);
+  }
+
+  /**
+   * Serializes a pegasus Data binding type to JSON.
+   * Valid types: DataMap, DataList, String, Number, Boolean, ByteString, Null
+   *
+   * The keys in a DataMap will be sorted in the order of fields in the
+   * corresponding schema.
+   *
+   * @param value the value to serialize to JSON.
+   * @return a JSON serialized string representation of the data value.
+   */
+  protected String toJson(Object value, JacksonDataTemplateCodec jsonCodec, DataSchema schema)
+      throws IOException
+  {
+    if (value instanceof DataMap || value instanceof DataList)
+    {
+      StringWriter out = new StringWriter(DEFAULT_JSON_CODEC_BUFFER_SIZE);
+      jsonCodec.writeDataTemplate(value, schema, out, true);
+      return out.toString();
+    }
+
+    return toPrimitiveValue(value);
+  }
+
+  private static String toPrimitiveValue(Object value)
```

Comment:
This naming isn't quite accurate. It's technically not transforming the data "to a primitive", it's transforming "from primitive data to primitive JSON data".

However, I think dividing the logic like this isn't totally necessary. I suggest packing this logic back into `toJson(value, codec)` and having `toJson(value, codec, schema)` just delegate back to that one for the primitive case.

Reply:
I missed the fact that `JacksonDataTemplateCodec` extends `JacksonDataCodec`, which makes this possible. Doing this now.